jsonComposer.go

```diff
@@ -0,0 +1,218 @@
+
+package gojsonsm
+
+import (
+	"fmt"
+)
+
+type jsonObjComposer struct {
+	// should have enough length
+	body []byte
+	// cursor
+	pos int
+	// if the newly composed json has atleast one field
+	atLeastOneFieldLeft bool
+}
+
+func (composer *jsonObjComposer) Write(data []byte) {
+	copy(composer.body[composer.pos:composer.pos+len(data)], data)
+	composer.pos += len(data)
+}
+
+func (composer *jsonObjComposer) Commit(stepBackAndClose bool) (int, bool) {
+	// close the JSON object after stepping back by one position explicity if requested
+	if stepBackAndClose && composer.pos-1 >= 0 {
+		if composer.body[composer.pos-1] != '{' {
+			composer.pos--
+		}
+		composer.body[composer.pos] = '}'
+		composer.pos++
+	}
+
+	// if the body is anything other than {}, we have atleast one field inside it
+	composer.atLeastOneFieldLeft = composer.pos > 2
+
+	return composer.pos, composer.atLeastOneFieldLeft
+}
+
+// Given a byte encoded json object - "src", a list of keys of items to remove from src - "remove",
+// the function removes the items from "src" and places them in "removed" and the remaining json object is stored in "dst"
+// returns (final length of dst, number of items removed, if there are any items left in dst at the end, error)
+// Caller has the ability to allocate memory for "dst" and "removed". If nil is passed, only then memory is allocated.
+func MatchAndRemoveItemsFromJsonObject(src []byte, remove []string, dst []byte, removed map[string][]byte) (int, int, bool, error) {
+	var removedLen int
+	if removed == nil {
+		removed = make(map[string][]byte)
+	}
+
+	if dst == nil {
+		dst = make([]byte, len(src))
+	}
+
+	composer := &jsonObjComposer{
+		body:                dst,
+		pos:                 0,
+		atLeastOneFieldLeft: false,
+	}
+
+	tokenizer := &jsonTokenizer{}
+	tokenizer.Reset(src)
+
+	var tknType tokenType
+	var tknLen, dstLen int
+	var tkn []byte
+	var err error
+	var atleastOneFieldLeft, removedLastItem bool
+	var depth int
+	for tknType != tknEnd {
+		tknType, tkn, tknLen, err = tokenizer.Step()
+		if err != nil {
+			return 0, 0, false, fmt.Errorf("error stepping to next token, src=%s, pos=%v, err=%v", src, tokenizer.Position(), err)
+		}
+
+		switch tknType {
+		case tknString:
+			// depth 1 strings are JSON object keys, need to check if it is the key to remove
+			if depth != 1 {
+				composer.Write(tkn)
+				continue
+			}
+		case tknObjectStart:
+			fallthrough
+		case tknArrayStart:
+			depth++
+			composer.Write(tkn)
+			continue
+		case tknObjectEnd:
+			fallthrough
+		case tknArrayEnd:
+			depth--
+			if depth < 0 {
+				return 0, 0, false, fmt.Errorf("invalid JSON object")
+			}
+			composer.Write(tkn)
+			continue
+		case tknEnd:
+			continue
+		default:
+			composer.Write(tkn)
+			continue
+		}
+
+		// strip off the quotes from the string
+		key := tkn[1 : tknLen-1]
+
+		matched := false
+		for _, keyToRemove := range remove {
+			if BytesEqualsString(key, keyToRemove) {
```

Comment:
For readability, do:
```
if !BytesEqualString {
continue
}
```

Then less indent is needed

Reply:
Done